ooni/scripts/ooniprobe_agent.py

```diff
@@ -0,0 +1,193 @@
+import errno
+import signal
+
+from twisted.scripts import twistd
+from twisted.python import usage
+
+from ooni.utils import log, is_process_running
+from ooni.settings import config
+from ooni.agent.agent import AgentService
+
+
+class StartOoniprobeAgentPlugin:
+    tapname = "ooniprobe"
+
+    def makeService(self, so):
+        return AgentService(config.advanced.webui_port)
+
+class OoniprobeTwistdConfig(twistd.ServerOptions):
+    subCommands = [
+        ("StartOoniprobeAgent", None, usage.Options, "ooniprobe agent")
+    ]
+
+class StartOptions(usage.Options):
+    pass
+
+class StopOptions(usage.Options):
+    pass
+
+class StatusOptions(usage.Options):
+    pass
+
+class RunOptions(usage.Options):
+    pass
+
+class AgentOptions(usage.Options):
+    subCommands = [
+        ['start', None, StartOptions, "Start the ooniprobe-agent in the "
+                                      "background"],
+        ['stop', None, StopOptions, "Stop the ooniprobe-agent"],
+        ['status', None, StatusOptions, "Show status of the ooniprobe-agent"],
+        ['run', None, RunOptions, "Run the ooniprobe-agent in the foreground"]
+    ]
+    def postOptions(self):
+        self.twistd_args = []
+
+def start_agent(options=None):
+    config.set_paths()
+    config.initialize_ooni_home()
+    config.read_config_file()
+
+    os.chdir(config.running_path)
+
+    # Since we are starting the logger below ourselves we make twistd log to
+    #  a null log observer
+    twistd_args = ['--logger', 'ooni.utils.log.ooniloggerNull',
+                   '--umask', '022']
+
+    twistd_config = OoniprobeTwistdConfig()
+    if options is not None:
+        twistd_args.extend(options.twistd_args)
+    twistd_args.append("StartOoniprobeAgent")
+    try:
+        twistd_config.parseOptions(twistd_args)
+    except usage.error, ue:
+        print("ooniprobe: usage error from twistd: {}\n".format(ue))
+    twistd_config.loadedPlugins = {
+        "StartOoniprobeAgent": StartOoniprobeAgentPlugin()
+    }
+
+    try:
+        get_running_pidfile()
+        print("Stop ooniprobe-agent before attempting to start it")
+        return 1
+    except NotRunning:
+        pass
+
+    print("Starting ooniprobe agent.")
+    WEB_UI_URL = "http://{0}:{1}".format(
+        config.advanced.webui_address, config.advanced.webui_port)
+    print("To view the GUI go to %s" % WEB_UI_URL)
+    log.start()
+    twistd.runApp(twistd_config)
+    return 0
+
+
+class NotRunning(RuntimeError):
+    pass
+
+def get_running_pidfile():
+    """
+    :return: This pid of the running ooniprobe-agent instance.
+    :raises: NotRunning if it's not running
+    """
+    running_pidfile = None
+    for pidfile in [config.system_pid_path, config.user_pid_path]:
+        if not os.path.exists(pidfile):
+            # Didn't find the pid_file
+            continue
+        pid = open(pidfile, "r").read()
+        pid = int(pid)
+        if is_process_running(pid):
+            running_pidfile = pidfile
```

Comment:
Yes fair point. Now that it has been refactored with a function to check if the process is running this can be rewritten as you suggest that is indeed much cleaner.